server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2639,22 +2639,23 @@ private int getMaxDataVolumesSupported(UserVmVO vm) {
         return maxDataVolumesSupported.intValue();
     }
 
-    private Long getDeviceId(long vmId, Long deviceId) {
+    private Long getDeviceId(UserVmVO vm, Long deviceId) {
         // allocate deviceId
-        List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        int maxDataVolumesSupported = getMaxDataVolumesSupported(vm);
+        List<VolumeVO> vols = _volsDao.findByInstance(vm.getId());
         if (deviceId != null) {
-            if (deviceId.longValue() > 15 || deviceId.longValue() == 3) {
-                throw new RuntimeException("deviceId should be 1,2,4-15");
+            if (deviceId.longValue() > maxDataVolumesSupported || deviceId.longValue() == 3) {
+                throw new RuntimeException("deviceId should be 1,2,4-" + maxDataVolumesSupported);
             }
             for (VolumeVO vol : vols) {
                 if (vol.getDeviceId().equals(deviceId)) {
-                    throw new RuntimeException("deviceId " + deviceId + " is used by vm" + vmId);
+                    throw new RuntimeException("deviceId " + deviceId + " is used by vm" + vm.getId());
                 }
             }
         } else {
             // allocate deviceId here
             List<String> devIds = new ArrayList<String>();
```

Comment:
Not part of your changes but... variable `devIds` should have type `List<Long>` instead of `List<String>`.
All that conversion from `int` to `String` and then converting from `String` to `long` seems unnecessary.
Should simply be able to do:
```
List<Long> devIds = new ArrayList<>();
for (long i = 1; i <= maxDataVolumesSupported; i++) {
    devIds.add(i);
}
devIds.remove(3L);
for (VolumeVO vol : vols) {
    devIds.remove(vol.getDeviceId());
}
if (devIds.isEmpty()) {
    throw new RuntimeException("every available deviceId already in use by vm " + vm.getId());
}
deviceId = devIds.iterator().next();
```
Note: my code above includes fixes to two other comments I made further down in the code.